imgix/urlbuilder.py

```diff
@@ -145,28 +175,89 @@ def create_srcset(self, path, params=None):
         if (has_width or (has_height and has_aspect_ratio)):
             return self._build_srcset_DPR(path, params)
         else:
-            return self._build_srcset_pairs(path, params)
+            targets = target_widths(start, stop, tol)
+            return self._build_srcset_pairs(path, params, targets)
 
-    def _build_srcset_pairs(self, path, params):
+    def _build_srcset_pairs(
+            self, path, params, targets=TARGET_WIDTHS):
         # prevents mutating the params dict
         srcset_params = dict(params)
         srcset_entries = []
 
-        for w in SRCSET_TARGET_WIDTHS:
+        for w in targets:
             srcset_params['w'] = w
             srcset_entries.append(self.create_url(path, srcset_params) +
                                   " " + str(w) + "w")
 
         return ",\n".join(srcset_entries)
 
-    def _build_srcset_DPR(self, path, params):
+    def _build_srcset_DPR(
+            self, path, params, targets=TARGET_RATIOS):
         # prevents mutating the params dict
         srcset_params = dict(params)
         srcset_entries = []
 
-        for dpr in SRCSET_DPR_TARGET_RATIOS:
+        for dpr in targets:
             srcset_params['dpr'] = dpr
             srcset_entries.append(self.create_url(path, srcset_params) +
                                   " " + str(dpr) + "x")
 
         return ",\n".join(srcset_entries)
+
+
+def target_widths(start=MIN_WIDTH, stop=MAX_WIDTH, tol=TOLERANCE):
```

Comment:
Yeah I considered this last point. I think we should expose more functionality. If this function is public it means everyone could use it (us and our users).

We actually do expose a target widths equivalent:
https://github.com/imgix/imgix-rb/blob/d5964b86ebb1c4611dab2aeb69703d28e68c264e/lib/imgix/path.rb#L145

https://github.com/imgix/imgix-rb/blob/d5964b86ebb1c4611dab2aeb69703d28e68c264e/lib/imgix/path.rb#L42

https://github.com/imgix/imgix-rb/blob/d5964b86ebb1c4611dab2aeb69703d28e68c264e/lib/imgix.rb#L21

<img width="647" alt="Screen Shot 2020-04-16 at 2 28 46 PM" src="https://user-images.githubusercontent.com/22702087/79508534-965f6680-7fee-11ea-9ddb-14f516050d28.png">

Also, I would like to discuss where `target_widths` should be defined, but in general it not relying on any class is a benefit in this case: more tools in the kit for users to use if they choose to (instead of trapping generic functionality within a class)

Noticed this as well:
https://github.com/imgix/imgix-rb/blob/d5964b86ebb1c4611dab2aeb69703d28e68c264e/lib/imgix.rb#L12

Reply:
Ah, sorry I didn't answer your first question, yes within the current implementation
>The interface to customize a srcset is provided through `create_srcset` so I'm curious what the reasoning would be for a user to want to call this by itself? 

Reasons:
* allow users to figure out what `target_widths` does without having to do everything else (i.e. define a builder, create a url etc.)
  * the function itself doesn't depend on any configuration other than what's passed in
  * moreover `create_srcset` only relies on what's passed in as well, however, one of the things that's passed is `self`

So, `create_srcset` is coupled with the class, but it doesn't need to be. It doesn't mutate its internal state. By this I mean it doesn't mutate any of these
```python
  self._domain = domain
  self._sign_key = sign_key
  self._use_https = use_https
  self._include_library_param = include_library_param
```
The class does call function that are internal to its implementation, i.e.
https://github.com/imgix/imgix-python/blob/2920949dd752f85c8f6b764f4092a0d0e66ba942/imgix/urlbuilder.py#L150

But in this case as well, the functions don't need access to internal state other than to call functions that also do not rely on modifying internal state.  If `target_widths` needed to know about or mutate some state that's internal to `class UrlBuilder`, e.g. `self.domain`, then wrapping it up in the class would seem natural. But that's not the case.

I just thought we could decouple the data configuration/representation from the logic/behaviors etc. Thoughts?

Reply:
Closing the loop:
- target_widths stays as a public method
- the point about decoupling logic is a good one, but it will need to be an ongoing conversation. For now it's not specifically concerned with these suggested changes.